connectors/sources/salesforce.py

```diff
@@ -398,6 +452,43 @@ async def _select_queryable_fields(self, sobject, fields):
         queryable_fields = sobject_fields.get(sobject, [])
         return [f for f in fields if f.lower() in queryable_fields]
 
+    def _parse_content_documents(self, record):
+        content_docs = []
+        content_links = record.get("ContentDocumentLinks", {}) or {}
+        content_links = content_links.get("records", []) or []
+        for content_link in content_links:
+            content_doc = content_link.get("ContentDocument")
+            if not content_doc:
+                continue
+
+            content_doc["linked_sobject_id"] = record.get("Id")
+            content_docs.append(content_doc)
+
+        return content_docs
+
+    def _combine_duplicate_content_docs(self, content_docs):
+        """Duplicate ContentDocuments may appear linked to multiple SObjects
+        Here we ensure that we don't download any duplicates while retaining links"""
+        grouped = {}
+
+        for content_doc in content_docs:
+            content_doc_id = content_doc["Id"]
+            if content_doc_id in grouped:
+                linked_id = content_doc["linked_sobject_id"]
+                linked_ids = grouped[content_doc_id]["linked_ids"]
+                if linked_id not in linked_ids:
+                    linked_ids.append(linked_id)
```

Comment:
Does it actually append to `grouped[content_doc_id]["linked_ids"]`?

I'm not 100% familiar with how ref variables work in Python

Reply:
it does!
Apparently in python `linked_ids` here isn't technically its own variable, its just a reference to that dict key. It threw me for a loop when I first realised.
(I think this is why calling `.copy()` is so important in python)